src/EFCore.Relational/Update/ModificationCommand.cs

```diff
@@ -341,25 +355,82 @@ private List<IColumnModification> GenerateColumnModifications()
+                columnMappings = storedProcedureMapping.ParameterMappings
+                    .Concat((IEnumerable<IColumnMappingBase>)storedProcedureMapping.ResultColumnMappings);
+
+                // Stored procedures may have an additional rows affected parameter, result column or return value, which does not have a
+                // property/column mapping but still needs to have be represented via a column modification.
+                var storedProcedure = storedProcedureMapping.StoredProcedure;
+
+                IColumnBase? rowsAffectedColumnBase = null;
+
+                if (storedProcedure.FindRowsAffectedParameter() is { } rowsAffectedParameter)
+                {
+                    rowsAffectedColumnBase = RowsAffectedColumn = rowsAffectedParameter.StoreParameter;
+                }
+                else if (storedProcedure.FindRowsAffectedResultColumn() is { } rowsAffectedResultColumn)
+                {
+                    rowsAffectedColumnBase = RowsAffectedColumn = rowsAffectedResultColumn.StoreResultColumn;
+                }
+                else if (storedProcedureMapping.StoreStoredProcedure.ReturnValue is { } rowsAffectedReturnValue)
+                {
+                    rowsAffectedColumnBase = rowsAffectedReturnValue;
+                }
+
+                if (rowsAffectedColumnBase is not null)
+                {
+                    Check.DebugAssert(_intTypeMapping is not null, "_intTypeMapping is not null");
 
-            foreach (var columnMapping in tableMapping.ColumnMappings)
+                    columnModifications.Add(CreateColumnModification(new ColumnModificationParameters(
+                        entry,
+                        property: null,
+                        rowsAffectedColumnBase,
+                        _generateParameterName!,
+                        _intTypeMapping!,
+                        valueIsRead: true,
+                        valueIsWrite: false,
+                        columnIsKey: false,
+                        columnIsCondition: false,
+                        _sensitiveLoggingEnabled)));
+                }
+            }
+
+            foreach (var columnMapping in columnMappings)
             {
                 var property = columnMapping.Property;
                 var column = columnMapping.Column;
+                var storedProcedureParameter = columnMapping is IStoredProcedureParameterMapping parameterMapping
+                    ? parameterMapping.Parameter
+                    : null;
                 var isKey = property.IsPrimaryKey();
-                var isCondition = !adding && (isKey || property.IsConcurrencyToken);
-                var readValue = state != EntityState.Deleted && entry.IsStoreGenerated(property);
+                var isCondition = !adding
+                    && (isKey || (property.IsConcurrencyToken && storedProcedureParameter is null or { ForOriginalValue: true }));
```

Comment:
```suggestion
                    && (isKey || ((property.IsConcurrencyToken && storedProcedureParameter is null) or storedProcedureParameter.ForOriginalValue == true));
```

We should support non-conditional original values

Reply:
Can you give me a scenario to codify in a test?

Reply:
We currently pass all property values even if they weren't modified, so a test scenario would be a sproc that compares the original and the current value and only updates it in the row if they are different

Reply:
OK, I've added support for this, but it currently feels hacky... The basic problem is that we [identify](https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Update/ColumnModification.cs#L95) the concept of OriginalValue with IsCondition. Here we're adding support for an original value that isn't a condition (in the concurrency token sense).

A more general fix could be to add state to IColumnModification to make it aware whether the original/current values (make Use{Original,Current}Value have a field instead of calculate from IsCondition/IsWrite), disconnecting the condition concept from original value etc. For now I've done the hackier way of continuing to use IsCondition=true to signal that the original value should be used; for sprocs it doesn't matter anyway where something is a condition or a write (no WHERE clause).

Reply:
Test: Original_and_current_value_on_non_concurrency_token